# Preparation

Preparation biasa dilakukan untuk mempersiapkan data sebelum masuk dalam tahap pemodelan. <br>
Berikut adalah tahapan yang akan dilalui pada data `SC_HW1_bank_data.csv` sebelum tahap pemodelan :
1. Import Library
2. Input Dataset
3. Preprocessing
4. Train-Test Split

## Import Library

In [ ]:
import pandas as pd
import numpy as np

## Input Dataset

In [ ]:
#Membaca data dan memasukkannya ke dalam bentuk Pandas Dataframe
df = pd.read_csv('https://raw.githubusercontent.com/Rietaros/kampus_merdeka/main/SC_HW1_bank_data.csv')

In [ ]:
#Jalankan code untuk mengecek nama kolom yang tersedia
df.columns

Index(['RowNumber', 'CustomerId', 'Geography', 'Gender', 'Age', 'Tenure',
       'Balance', 'NumOfProducts', 'HasCrCard', 'IsActiveMember',
       'EstimatedSalary', 'Exited'],
      dtype='object')

In [ ]:
#Hilangkan kolom yang dirasa tidak relevan dengan model (contoh: ID). None dapat diisi dengan nama-nama kolom yang akan digunakan.
#Contoh df = df[['X1','X2', 'Y']].copy()

#START CODE
df = df[['Geography', 'Gender', 'Age', 'Tenure', 'Balance', 'NumOfProducts', 'HasCrCard', 'IsActiveMember', 'EstimatedSalary', 'Exited']].copy()

In [ ]:
data_object = df.select_dtypes(include = ["object"]).columns

for column in data_object:
  df[column] = df[column].astype("category")

df.dtypes

Geography          category
Gender             category
Age                   int64
Tenure                int64
Balance             float64
NumOfProducts         int64
HasCrCard             int64
IsActiveMember        int64
EstimatedSalary     float64
Exited                int64
dtype: object

## Preprocessing

In [ ]:
#Lakukan One-Hot Encoder untuk data categorical, dengan fungsi pandas get_dummies

#START CODE
df = pd.get_dummies(df, columns=['Geography', 'Gender'])

In [ ]:
#Pisahkan mana X (feature) dengan Y,
#Y adalah kolom "Exited"

#START CODE
X = df.drop(columns=['Exited'])
y = df[['Exited']]

In [ ]:
#Lakukan Scaler dan/atau Noermalisasi jika diperlukan
from sklearn.preprocessing import MinMaxScaler

#START CODE
scaler = MinMaxScaler()
X_transform = scaler.fit_transform(X)

In [ ]:
#Ini digunakan jika dilakukan scaler/Normalisas. Jika tidak, code ini bisa dilewat dan diganti dengan code yang ada di dalam komen
X_transform = pd.DataFrame(X_transform, columns = X.columns)
#X_transform = X.copy()

## Train-Test Split

In [ ]:
#Split menjadi train dan test dengan test_size 25%
#Tidak perlu mengubah code ini

from sklearn.model_selection import train_test_split
X_train,X_test,y_train,y_test = train_test_split(X_transform,y,test_size = 0.25,random_state = 123)

# Modeling

## Model1

### Soal :
Jelaskan secara Singkat Model pertama yang digunakan!

In [ ]:
#Pilih salah satu metode Machine Leaarning
#Model Machine Learning dapat dipanggil terlebih dahulu melalui library yang digunakan. Gunakan library scikit learn seperti pada contoh

#Contoh pemanggilan library dan penggunaannya dalam model

#from sklearn.linear_model import LogisticRegression
#model1 = LogisticRegression()
#params = { "tol": [0.1,0.01,0.001], 'C':[0.5,1.0,1.5,2.0]}


#START CODE
from sklearn.linear_model import LogisticRegression
from sklearn.model_selection import GridSearchCV

model1 = LogisticRegression()
params = {
    'tol': [0.1, 0.01, 0.001],
    'C': [0.5, 1.0, 1.5, 2.0]
}

grid = GridSearchCV(
    estimator=model1,
    param_grid=params,
    scoring='accuracy',
    n_jobs=10,
    cv=10
)

grid.fit(X_train, y_train)

best_params = grid.best_params_
print("Parameter terbaik:", best_params)

Parameter terbaik: {'C': 0.5, 'tol': 0.1}


/usr/local/lib/python3.10/dist-packages/sklearn/utils/validation.py:1143: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().
  y = column_or_1d(y, warn=True)


In [ ]:
#lakukan evaluasi dengan beberapa maetric di bawah ini
from sklearn.metrics import classification_report
from sklearn.metrics import accuracy_score
from sklearn.metrics import confusion_matrix

y_pred = grid.predict(X_test)

print("Classification Report:")
print(classification_report(y_test, y_pred))

print("\nConfusion Matrix:")
print(confusion_matrix(y_test, y_pred))

accuracy = accuracy_score(y_test, y_pred)
print("\nAccuracy:", accuracy)

Classification Report:
              precision    recall  f1-score   support

           0       0.82      0.97      0.89      1983
           1       0.66      0.20      0.31       517

    accuracy                           0.81      2500
   macro avg       0.74      0.59      0.60      2500
weighted avg       0.79      0.81      0.77      2500


Confusion Matrix:
[[1929   54]
 [ 412  105]]

Accuracy: 0.8136


## Model2
### Soal :
Jelaskan secara Singkat Model ke-2 yang digunakan!

In [ ]:
#Pilih salah satu metode Machine Leaarning
#Model Machine Learning dapat dipanggil terlebih dahulu melalui library yang digunakan. Gunakan library scikit learn seperti pada contoh

#Contoh pemanggilan library dan penggunaannya dalam model

#from sklearn.linear_model import LogisticRegression
#model1 = LogisticRegression()
#params = { "tol": [0.1,0.01,0.001], 'C':[0.5,1.0,1.5,2.0]}


#START CODE
from sklearn.neighbors import KNeighborsClassifier
from sklearn.model_selection import GridSearchCV

model2 = KNeighborsClassifier()

params = {
    'n_neighbors': [3, 5, 7],
    'weights': ['uniform', 'distance'],
    'p': [1, 2]
}

grid = GridSearchCV(
    estimator=model2,
    param_grid=params,
    scoring='accuracy',
    n_jobs=10,
    cv=10
)

grid.fit(X_train, y_train)

best_params = grid.best_params_
print("Parameter terbaik:", best_params)

Parameter terbaik: {'n_neighbors': 7, 'p': 1, 'weights': 'distance'}


/usr/local/lib/python3.10/dist-packages/sklearn/neighbors/_classification.py:215: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples,), for example using ravel().
  return self._fit(X, y)


In [ ]:
#lakukan evaluasi
from sklearn.metrics import classification_report
from sklearn.metrics import accuracy_score
from sklearn.metrics import confusion_matrix

y_pred = grid.predict(X_test)

print("Classification Report:")
print(classification_report(y_test, y_pred))

print("\nConfusion Matrix:")
print(confusion_matrix(y_test, y_pred))

accuracy = accuracy_score(y_test, y_pred)
print("\nAccuracy:", accuracy)

Classification Report:
              precision    recall  f1-score   support

           0       0.85      0.95      0.89      1983
           1       0.63      0.34      0.45       517

    accuracy                           0.82      2500
   macro avg       0.74      0.65      0.67      2500
weighted avg       0.80      0.82      0.80      2500


Confusion Matrix:
[[1879  104]
 [ 339  178]]

Accuracy: 0.8228


## Model3
### Soal :
Jelaskan secara Singkat Model ke-3 yang digunakan!

In [ ]:
#Pilih salah satu metode Machine Leaarning
#Model Machine Learning dapat dipanggil terlebih dahulu melalui library yang digunakan. Gunakan library scikit learn seperti pada contoh

#Contoh pemanggilan library dan penggunaannya dalam model

#from sklearn.linear_model import LogisticRegression
#model1 = LogisticRegression()
#params = { "tol": [0.1,0.01,0.001], 'C':[0.5,1.0,1.5,2.0]}


#START CODE
from sklearn.ensemble import RandomForestClassifier
from sklearn.model_selection import GridSearchCV

model3 = RandomForestClassifier()

params = {
    'n_estimators': [50, 100, 200, 300],  # Jumlah pohon dalam ensemble
    'max_depth': [5, 10, 20, 30],  # Kedalaman maksimum setiap pohon
    'min_samples_split': [2, 5, 10],  # Jumlah sampel minimum untuk membagi simpul internal
    'min_samples_leaf': [1, 2, 4, 5, 10, 20],  # Jumlah sampel minimum di setiap daun pohon
}

grid = GridSearchCV(
    estimator=model3,
    param_grid=params,
    scoring='accuracy',
    n_jobs=10,
    cv=10
)

grid.fit(X_train, y_train)

best_params = grid.best_params_
print("Parameter terbaik:", best_params)

/usr/local/lib/python3.10/dist-packages/sklearn/model_selection/_search.py:909: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples,), for example using ravel().
  self.best_estimator_.fit(X, y, **fit_params)


Parameter terbaik: {'max_depth': 10, 'min_samples_leaf': 5, 'min_samples_split': 2, 'n_estimators': 200}


In [ ]:
#lakukan evaluasi
from sklearn.metrics import classification_report
from sklearn.metrics import accuracy_score
from sklearn.metrics import confusion_matrix

y_pred = grid.predict(X_test)

print("Classification Report:")
print(classification_report(y_test, y_pred))

print("\nConfusion Matrix:")
print(confusion_matrix(y_test, y_pred))

accuracy = accuracy_score(y_test, y_pred)
print("\nAccuracy:", accuracy)

Classification Report:
              precision    recall  f1-score   support

           0       0.87      0.97      0.92      1983
           1       0.78      0.44      0.56       517

    accuracy                           0.86      2500
   macro avg       0.83      0.70      0.74      2500
weighted avg       0.85      0.86      0.84      2500


Confusion Matrix:
[[1921   62]
 [ 291  226]]

Accuracy: 0.8588


## Tarik Kesimpulan Model Mana yang terbaik beserta alasannya

Dari ketiga model di atas di dapat akurasi sebagai berikut:
1. Logistic Regression: 0.8136
2. K-Nearest Neighbors (K-NN): 0.8228
3. Random Forest: 0.8588

Oleh karena itu, di antara ketiga model yang digunakan, Random Forest dianggap sebagai pilihan terbaik. karena random forest memilii akurasi yang lebih tinggi dibandingkan dengan model lainnya. Model ini memiliki keunggulan dalam menangani masalah non-linear dan memiliki kemampuan ensemble yang dapat meningkatkan akurasi. Selain itu, Random Forest biasanya lebih tahan terhadap overfitting daripada model lainnya. Meskipun begitu, perlu diingat bahwa penggunaan Random Forest dengan jumlah tree yang sangat besar dapat memakan waktu yang cukup lama, sehingga perlu dilakukan tuning hyperparameter dengan hati-hati.